rbd/snapshot.go

```diff
@@ -147,6 +148,34 @@ func (snapshot *Snapshot) IsProtected() (bool, error) {
 	return cIsProtected != 0, nil
 }
 
+// Rename a snapshot.
+// PREVIEW
+//
+// Implements:
+// 	int rbd_snap_rename(rbd_image_t image, const char *snapname,
+//				 const char* dstsnapsname);
+func (snapshot *Snapshot) Rename(destName string) error {
+	if err := snapshot.validate(imageNeedsIOContext | imageIsOpen | imageNeedsName | snapshotNeedsName); err != nil {
+		return err
+	}
+
+	cSrcName := C.CString(snapshot.name)
+	cDestName := C.CString(destName)
+	defer C.free(unsafe.Pointer(cSrcName))
+	defer C.free(unsafe.Pointer(cDestName))
+
+	err := rbdError(C.rbd_snap_rename(snapshot.image.image, cSrcName, cDestName))
+	if err != 0 {
+		if err == -C.EEXIST {
```

Comment:
Why do you want to "hide" the EEXIST error condition? Callers could not distinguish between this and a success call.  I strongly suggest dropping this and if you want to, you can ignore EEXIST via the ErrorCode function in your code that calls Rename (if you are not clear on how this works, please just ask).

Reply:
I think I am misunderstanding something here. I initially considered using `err.ErrorCode()` to filter out EEXIST at the client level, but was foiled by fact that `rbdError` is unexported, so this isn't possible:
```
	if err := snap.Rename(...); err != nil {
		errno := err.(rbd.RbdError).ErrorCode()
		if errno == EEXIST {
			return nil
		}
	}

``` 
Is there a better approach?

Reply:
> > Once this is satisfactory, I would love to backport to v0.13 and v1.14.
> 
> I'm sorry to disappoint but go-ceph doesn't do backports. Unlike ceph itself, each release is effectively a checkpoint in time. We do releases every other month. If you'd like to discuss this further, feel free to start a new github discussion on this topic.

Do you have a recommended workaround for how to get a version of go-ceph that's old enough to be compatible with a Nautilus Ceph cluster but also contains patches like this one?

Reply:
> I think I am misunderstanding something here. I initially considered using `err.ErrorCode()` to filter out EEXIST at the client level, but was foiled by fact that `rbdError` is unexported, so this isn't possible:
> 
> ```
> 	if err := snap.Rename(...); err != nil {
> 		errno := err.(rbd.RbdError).ErrorCode()
> 		if errno == EEXIST {
> 			return nil
> 		}
> 	}
> ```
> 
> Is there a better approach?

Indeed, there is. We should probably do a better job of advertising this, but you don't need to have rbd exported to use the ErrorCode method. Briefly: 
```golang
cephErr, ok := err.(interface{ ErrorCode() int })
```
OR
```golang
type errorWithCode interface {
    ErrorCode() int
}
var ec interface{ ErrorCode() int }
ok := errors.As(err, &ec)
```

These examples are pulled from our test code. look in cephfs/errors_test.go  and cephfs/admin/mirror_workflow_test.go for more fleshed out examples of the above. (Don't let the cephfs context turn you away. The approach will work the same for rbd)

Reply:
Thank you for the explanation!